### ISSUE TITLE
ModelAveragingRegressor: one non-finite member score makes every weight NaN and SignalOptimizer stale-fills the forecast silently

### ISSUE THREAD

**Issue by fscereda** [macrosynergy/macrosynergy#2757]
## Summary

`ModelAveragingRegressor._compute_weights` (`macrosynergy/learning/model_selection/model_averaging.py`, v1.8.1 and `develop`) turns **every** ensemble weight into NaN as soon as **one** member's cross-validation score is non-finite. `predict()` then returns NaN for the whole ensemble, and `SignalOptimizer` forward-fills the previous retrain's forecast over it (`forecasts_df.groupby(level=0).ffill()` in `signal_optimizer.py`), so the signal keeps trading a stale value with no warning or error.

Two related cases share the root cause and are already marked `# TODO: deal with zero later` in the source: a single-member ensemble (spread 0) and members with exactly tied scores (spread 0) both divide by zero.

## How it happens

1. A member's predictions are constant in a validation fold. The common case is a Lasso whose penalty is large enough to zero every coefficient on that fold; a `DummyRegressor` reproduces it deterministically.
2. Any correlation-based scorer (`correlation_coefficient`, `neg_mean_abs_corr`) is undefined for a constant prediction and returns NaN for that member.
3. `_compute_weights` does `spread = np.max(all_scores) - np.min(all_scores)` (or the std / mad / iqr variants). With one NaN in `all_scores`, `spread` is NaN.
4. `scaled_scores = all_scores / spread` is NaN for all members, so `weights` is NaN for all members, including those that scored perfectly well.
5. `predict` returns `predictions @ weights` = NaN. In `SignalOptimizer`, the NaN forecast is forward-filled and dropped silently.

## Observed impact

Real use: eleven asset-class sleeves, a ten-member Ridge/Lasso ensemble (Ridge alpha 1 to 10,000, Lasso alpha 0.01 to 1), `make_scorer(correlation_coefficient)`, `ExpandingKFoldPanelSplit(n_splits=3)`, quarterly retraining via `SignalOptimizer(...).calculate_predictions(inner_splitters=None, scorers=None)`. Member weights were NaN at **96 of 1,000** retraining dates, all caused by the Lasso alpha 1 member; one sleeve was affected at 51 of its 92 retrains. The stale-filled book had a *higher* information ratio than the correct one, so the defect would have been adopted as the best design had the weights not been inspected.

## Minimal reproduction

```python
import numpy as np, pandas as pd
from sklearn.dummy import DummyRegressor
from sklearn.linear_model import Ridge
from sklearn.metrics import make_scorer
from macrosynergy.learning import ModelAveragingRegressor, ExpandingKFoldPanelSplit, correlation_coefficient

rng = np.random.default_rng(0)
idx = pd.MultiIndex.from_product([["A", "B", "C"], pd.bdate_range("2010-01-29", periods=120, freq="BME")],
                                 names=["cid", "real_date"])
X = pd.DataFrame(rng.normal(size=(len(idx), 4)), index=idx, columns=list("wxyz"))
y = pd.Series(X.values @ np.array([0.5, 0.2, -0.2, -0.5]) + rng.normal(scale=0.5, size=len(idx)), index=idx)

m = ModelAveragingRegressor(
    estimators=[("ridge_1", Ridge(alpha=1.0), {}),
                ("ridge_100", Ridge(alpha=100.0), {}),
                ("constant", DummyRegressor(strategy="constant", constant=0.0), {})],
    scoring=make_scorer(correlation_coefficient),
    cv=ExpandingKFoldPanelSplit(n_splits=3),
).fit(X, y)
print(m.cv_scores_)   # {'ridge_1': 0.6..., 'ridge_100': 0.6..., 'constant': nan}
print(m.weights_)     # all three NaN
print(np.isnan(m.predict(X)).all())   # True
```

Also: `ModelAveragingRegressor(estimators=[("ridge", Ridge(), {})], ...)` gives weight NaN (spread 0 → 0/0), and two members with identical scores likewise.

## Suggested fix (for the owners to implement as they see fit)

Exclude members with non-finite scores from the softmax (weight 0, with a warning), compute the spread over finite scores only, and use the softmax limit (equal weights over the finite members) when the spread is zero or non-finite. Keep the `min_weight` quantiles on the finite members' weights so excluded zeros do not drag the floor down. Behaviour for all-finite, non-tied scores is unchanged.

```python
def _compute_weights(self, cv_scores, temperature, min_weight):
    all_scores = np.array([cv_scores[name] for name in self.model_names_], dtype=float)

    finite = np.isfinite(all_scores)
    if not finite.any():
        warnings.warn("No estimator produced a finite cross-validation score; "
                      "falling back to equal weights.", RuntimeWarning)
        return dict(zip(self.model_names_, np.full(len(all_scores), 1.0 / len(all_scores))))
    if not finite.all():
        excluded = [n for n, f in zip(self.model_names_, finite) if not f]
        warnings.warn(f"Estimators with a non-finite cross-validation score receive zero "
                      f"weight: {excluded}.", RuntimeWarning)
    scores = all_scores[finite]

    if temperature == "max-min":
        spread = np.max(scores) - np.min(scores)
    elif temperature == "std":
        spread = np.std(scores)
    elif temperature == "mad":
        spread = np.median(np.abs(scores - np.median(scores)))
    elif temperature == "iqr":
        spread = np.percentile(scores, 75) - np.percentile(scores, 25)
    else:
        spread = float(temperature)

    if not np.isfinite(spread) or spread <= 0:
        softmax = np.full(len(scores), 1.0 / len(scores))   # single member or exact ties
    else:
        scaled = scores / spread
        scaled = scaled - np.max(scaled)
        softmax = np.exp(scaled)
        softmax = softmax / softmax.sum()

    weights = np.zeros(len(all_scores))
    weights[finite] = softmax

    if isinstance(min_weight, str):
        ref = weights[finite]
        # ... existing "mean" / "median" / "lq" / "uq" / "lb" / "ub" branches, computed on ref ...

    # ... existing min_weight > 0 renormalisation unchanged ...
    return dict(zip(self.model_names_, adjusted_weights))
```

Tests worth adding: a constant-prediction member yields finite weights with that member at 0 and the rest summing to 1, and finite predictions; a single member yields weight 1 and reproduces that member's predictions; tied members yield equal weights; an all-non-finite ensemble falls back to equal weights; `min_weight="median"` ignores excluded members. A draft of these tests exists and can be shared on request.

## Possibly related

`correlation_coefficient` returning 0 rather than NaN for a constant prediction would make correlation scorers well-defined everywhere; that is a separate judgement call.

## Environment

macrosynergy 1.8.1 (also present on `develop` at bb353927), scikit-learn 1.8.0, numpy 2.4.6, Python 3.11.
